sw/host/opentitanlib/src/debug/dmi.rs

```diff
@@ -0,0 +1,332 @@
+    /// DMI registers on its own, which will interfere with raw DMI operations.
+    pub fn new(mut openocd: OpenOcd, tap: &str) -> Result<Self> {
+        openocd.irscan(tap, DTMCS)?;
+        let res = openocd.drscan(tap, 32, DTMCS_DMIRESET_MASK)?;
+        let abits = (res & DTMCS_ABITS_MASK) >> DTMCS_ABITS_SHIFT;
+
+        openocd.irscan(tap, DMI)?;
+        Ok(Self {
+            openocd,
+            tap: tap.to_owned(),
+            abits,
+        })
+    }
+
+    fn dmi_op(&mut self, op: u64) -> Result<u64> {
+        let res = self
+            .openocd
+            .drscan(&self.tap, self.abits + DMI_ADDRESS_SHIFT, op)?;
+
+        // We just scanned into the DMI register, so the scanned result should be empty.
+        ensure!(res == 0, "Unexpected DMI initial response {res:#x}");
+
+        // Run the DMI operation.
+        self.openocd.execute("runtest 10")?;
+
+        // Read the result.
+        let res = self
+            .openocd
+            .drscan(&self.tap, self.abits + DMI_ADDRESS_SHIFT, 0)?;
+        ensure!(res & 3 == 0, "DMI operation failed with {res:#x}");
+
+        // Double check the address matches.
+        ensure!(
+            res >> DMI_ADDRESS_SHIFT == op >> DMI_ADDRESS_SHIFT,
+            "DMI operation address mismatch {res:#x}"
+        );
+
+        Ok(res)
+    }
+}
+
+impl Dmi for OpenOcdDmi {
+    fn dmi_read(&mut self, addr: u32) -> Result<u32> {
+        let output = (self.dmi_op((addr as u64) << DMI_ADDRESS_SHIFT | DMI_OP_READ)?
+            >> DMI_DATA_SHIFT) as u32;
+        log::info!("DMI read {:#x} -> {:#x}", addr, output);
+        Ok(output)
+    }
+
+    fn dmi_write(&mut self, addr: u32, value: u32) -> Result<()> {
+        self.dmi_op(
+            (addr as u64) << DMI_ADDRESS_SHIFT | (value as u64) << DMI_DATA_SHIFT | DMI_OP_WRITE,
+        )?;
+        log::info!("DMI write {:#x} <- {:#x}", addr, value);
+        Ok(())
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum DmiError {
+    #[error("Hart does not exist")]
+    Nonexistent,
+    #[error("Hart is not currently available")]
+    Unavailable,
+    #[error("Timeout waiting for hart to halt")]
+    WaitTimeout,
+}
+
+/// A debugger that communicates with the target via RISC-V Debug Module Interface (DMI).
+pub struct DmiDebugger<D> {
+    dmi: D,
+}
+
+impl<D> Deref for DmiDebugger<D> {
+    type Target = D;
+
+    fn deref(&self) -> &Self::Target {
+        &self.dmi
+    }
+}
+
+impl<D> DerefMut for DmiDebugger<D> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.dmi
+    }
+}
+
+impl<D: Dmi> DmiDebugger<D> {
+    pub fn new(dmi: D) -> Self {
+        Self { dmi }
+    }
+
+    /// Selects a hart to debug.
+    pub fn select_hart(&mut self, hartid: u32) -> Result<DmiHart<'_, D>> {
+        // The hart selection is up to 20 bits.
+        if hartid >= (1 << 20) {
+            bail!("Invalid hartid: {hartid}");
+        }
+        let hart_select = 0 * DMCONTROL_HASEL_MASK
```

Comment:
The `0 * DMCONTROL_HASEL_MASK` is a bit strange.

Reply:
I use this to mean that the bit HASEL is set to 0. Do you prefer `0 << DMCONTROL_HASEL_SHIFT`, `0 /* HASEL */`, or just leave it out?

Reply:
Either one of the first two is fine, it's more the `0 *` which I found strange.